R/layouts.R

```diff
@@ -251,3 +251,63 @@ split_layout <- function(..., cell_widths = NULL, cell_args = "", style = NULL){
 splitLayout <- function(..., cellWidths = NULL, cellArgs = "", style = NULL) {
   split_layout(..., cell_widths = cellWidths, cell_args = cellArgs, style = style)
 }
+
+#' Flow layout
+#'
+#' Lays out elements in a left-to-right, top-to-bottom arrangement.
+#' The elements on a given row will be top-aligned with each other.
+#'
+#' The width of the elements and spacing between them is configurable.
+#' Lengths can be given as numeric values (interpreted as pixels)
+#' or character values (interpreted as CSS lengths).
+#' With the default settings this layout closely resembles the `flowLayout`
+#' from Shiny.
+#'
+#' @param ...
+#' Unnamed arguments will become child elements of the layout.
+#' Named arguments will become HTML attributes on the outermost tag.
+#' @param cell_args
+#' Any additional attributes that should be used for each cell of the layout.
+#' @param cell_width The width of the cells.
+#' @param column_gap The spacing between columns.
+#' @param row_gap The spacing between rows.
+#'
+#' @md
+#' @export
+#' @rdname flow_layout
+#'
+#' @examples
+#' if (interactive()) {
+#'   ui <- semanticPage(
+#'     flow_layout(
+#'       numericInput("rows", "How many rows?", 5),
+#'       selectInput("letter", "Which letter?", LETTERS),
+#'       sliderInput("value", "What value?", 0, 100, 50)
+#'     )
+#'   )
+#'   shinyApp(ui, server = function(input, output) {})
+#' }
+flow_layout <- function(..., cell_args = list(), cell_width = "208px", column_gap = "12px", row_gap = "0px") {
```

Comment:
why `cell_width` has some arbitrary values? if anything shouldn't these be percentages to account for different size sizes?

Reply:
These are the same arbitrary values used by Shiny ([code](https://github.com/rstudio/shiny/blob/24f3c20f2642fb932d5f21958f58e0a93fa033f2/inst/www/shared/shiny.css#L277)). I explained it in the doc:
> With the default settings this layout closely resembles the `flowLayout` from Shiny.

Using percentages would imply a fixed number of elements per row and this layout is not well suited for that purpose. User would have to take into account the space required by the gaps between the columns, e.g. using `cell_width = "33%"` would most likely lead to just two columns taking about 70% of the width.

Please let me know if you'd still like to see different default values here.

Reply:
No, you're right, if that's how Shiny does it then it's okay!